### PR TITLE
feat: config drift

### PR DESCRIPTION
**Pull request by arunjaindev**
# Description
feat: diff between live and generated manifest

Fixes https://github.com/devtron-labs/sprint-tasks/issues/757

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/v2/appDetails/appDetails.type.ts [hunk 1]
```diff
@@ -470,6 +471,7 @@ export interface ManifestViewRefType {
         manifest: string
         activeManifestEditorData: string
         modifiedManifest: string
+        normalizedLiveManifest: string
```
Comment (eshankvaish) [inline src/components/v2/appDetails/appDetails.type.ts:474]: add js doc

File: src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx [hunk 2]
```diff
@@ -418,16 +418,24 @@ const NodeComponent = ({
...
+                                    {(node?.info || node?.health?.status || node?.status) && (
+                                        <span
+                                            data-testid="node-resource-status"
+                                            className={`app-summary__status-name f-${(
+                                                node?.status ||
+                                                node?.health?.status ||
+                                                ''
+                                            ).toLowerCase()}`}
+                                        >
+                                            {getNodeStatus(node)}
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx:430]: let's create a variable for getNodeStatus and reuse

File: src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx [hunk 3]
```diff
@@ -418,16 +418,24 @@ const NodeComponent = ({
...
+                                        </span>
+                                    )}
+                                    {node.hasDrift && (
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx:433]:
```suggestion
                                    {node?.hasDrift && (
```

File: src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx [hunk 4]
```diff
@@ -418,16 +418,24 @@ const NodeComponent = ({
...
+                                        <>
+                                            {node?.health?.status && <span className="dc__bullet mw-4" />}
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx:435]:
```suggestion
                                            {node.health?.status && <span className="dc__bullet mw-4" />}
```

Comment (eshankvaish) [reply src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx:435]: We can move this part to ent and pass down the node to it

File: src/components/v2/appDetails/k8Resource/nodeDetail/nodeDetail.api.ts [hunk 5]
```diff
@@ -52,6 +55,9 @@ export const getManifestResource = (
     const requestData = isResourceBrowserView
         ? createResourceRequestBody(selectedResource)
         : createBody(ad, podName, nodeType)
+    if (getDesiredAndLiveManifest && ad.appType === AppType.DEVTRON_APP && !isResourceBrowserView) {
+        return getDesiredAndLiveManifest(requestData)
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeDetail/nodeDetail.api.ts:59]: pass down signal

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetail.component.tsx [hunk 6]
```diff
@@ -435,7 +442,7 @@ const NodeDetailComponent = ({
                 className={`w-100 pr-20 pl-20 bcn-0 flex dc__border-bottom dc__content-space ${!isResourceBrowserView ? 'node-detail__sticky' : ''}`}
             >
                 <div className="flex left">
-                    <div data-testid="app-resource-containor-header" className="flex left">
+                    <div data-testid="app-resource-container-header" className="flex left">
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetail.component.tsx:445]: undo

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/Manifest.component.tsx [hunk 7]
```diff
@@ -119,6 +121,12 @@ const ManifestComponent = ({
         }
     }
 
+    const _selectedResource = isResourceBrowserView
+        ? selectedResource
+        : appDetails.resourceTree.nodes.filter(
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/Manifest.component.tsx:126]: check for null checks

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/Manifest.component.tsx [hunk 8]
```diff
@@ -195,8 +199,15 @@ const ManifestComponent = ({
                 ])
                     .then((response) => {
                         setSecretViewAccess(response[0]?.result?.secretViewAccess || false)
-                        const _manifest = JSON.stringify(response[0]?.result?.manifestResponse?.manifest || '')
-                        setDesiredManifest(response[1]?.result?.manifest || '')
+                        const _manifest = JSON.stringify(
+                            response[0]?.result?.manifestResponse?.manifest || response[0]?.result?.liveState || '',
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/Manifest.component.tsx:203]: reverse the condition as discussed

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/Manifest.component.tsx [hunk 9]
```diff
@@ -514,6 +530,21 @@ const ManifestComponent = ({
...
+                                !showManifestCompareView && (
+                                    <div className="flexbox bcy-8 cn-0 px-20 py-4 dc__gap-8">
+                                        <span>Out of sync from desired manifest.</span>
+                                        <button
+                                            className="dc__unset-button-styles dc__underline"
+                                            onClick={handleDesiredManifestOpen}
+                                        >
+                                            Check configuration drift
+                                        </button>
+                                    </div>
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/Manifest.component.tsx:546]: move to FE lib

File: src/components/app/list-new/list.utils.ts [hunk 10]
```diff
@@ -148,7 +148,7 @@ export const useFilterOptions = ({
         () =>
             clusterGroupedEnvOptions?.map((clusterItem) => ({
                 label: getFormattedFilterValue(AppListUrlFilters.cluster, clusterItem.label),
-                options: clusterItem.options,
+                options: clusterItem.options.sort((a, b) => stringComparatorBySortOrder(a.label, b.label)),
```
Comment (eshankvaish) [inline src/components/app/list-new/list.utils.ts:151]: add null check

File: src/components/app/details/appDetails/SourceInfo.tsx [hunk 11]
```diff
@@ -170,6 +170,15 @@ export const SourceInfo = ({
...
                 )}
+                {appDetails?.resourceTree && ConfigSyncStatusButton && (
+                    <div className="pl-8">
+                        <ConfigSyncStatusButton
+                            areConfigurationsDrifted={appDetails.resourceTree.hasDrift}
+                            appId={appDetails?.appId}
+                            envId={envId}
+                        />
+                    </div>
+                )}
```
Comment (eshankvaish) [inline src/components/app/details/appDetails/SourceInfo.tsx:181]: or we can can not of isVirtual
```suggestion
                {appDetails?.resourceTree && !isIsolatedEnv && ConfigSyncStatusButton && (
                    <div className="pl-8">
                        <ConfigSyncStatusButton
                            areConfigurationsDrifted={appDetails.resourceTree.hasDrift}
                            appId={appDetails?.appId}
                            envId={envId}
                        />
                    </div>
                )}
```

File: src/components/app/details/appDetails/SourceInfo.tsx [hunk 12]
```diff
@@ -170,6 +170,15 @@ export const SourceInfo = ({
                         </div>
                     </Tooltip>
                 )}
+                {appDetails?.resourceTree && ConfigSyncStatusButton && (
+                    <div className="pl-8">
+                        <ConfigSyncStatusButton
+                            areConfigurationsDrifted={appDetails.resourceTree.hasDrift}
+                            appId={appDetails?.appId}
```
Comment (eshankvaish) [inline src/components/app/details/appDetails/SourceInfo.tsx:177]:
```suggestion
                            appId={appDetails.appId}
```

File: src/components/app/details/appDetails/AppDetails.tsx [hunk 13]
```diff
@@ -773,7 +781,9 @@ export const Details: React.FC<DetailsType> = ({
             ) : (
                 renderAppDetails()
             )}
-            {detailedStatus && <AppStatusDetailModal close={hideAppDetailsStatus} showAppStatusMessage={false} />}
+            {detailedStatus && (
+                <AppStatusDetailModal close={hideAppDetailsStatus} showAppStatusMessage={false} showConfigDriftInfo />
```
Comment (eshankvaish) [inline src/components/app/details/appDetails/AppDetails.tsx:785]: showConfigDriftInfo should be based of fe-lib and to be verified for isolated airgapped

File: src/components/app/details/appDetails/AppDetails.tsx [hunk 14]
```diff
@@ -815,6 +825,7 @@ export const Details: React.FC<DetailsType> = ({
                     isVirtualEnvironment={isVirtualEnvRef.current}
                 />
             }
+            {ConfigDriftModalRoute && <ConfigDriftModalRoute path={path} />}
```
Comment (eshankvaish) [inline src/components/app/details/appDetails/AppDetails.tsx:828]: route should not be there for virtual

**Review by eshankvaish — CHANGES_REQUESTED** [review]

**Review by eshankvaish — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**